spec_validator.go

```diff
@@ -0,0 +1,144 @@
+/**
+ *  Copyright (c) 2022  Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package wrp
+
+import (
+	"errors"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/google/uuid"
+	"go.uber.org/multierr"
+)
+
+const (
+	serialPrefix = "serial"
+	uuidPrefix   = "uuid"
+	eventPrefix  = "event"
+	dnsPrefix    = "dns"
+)
+
+var (
+	ErrorInvalidMessageEncoding = errors.New("invalid message encoding")
+	ErrorInvalidMessageType     = errors.New("invalid message type")
+	ErrorInvalidLocator         = errors.New("invalid locator")
+	ErrorInvalidSource          = errors.New("invalid Source name")
+	ErrorInvalidDestination     = errors.New("invalid Destination name")
+)
+
+// locatorPattern is the precompiled regular expression that all source and dest locators must match.
+// Matching is partial, as everything after the authority (ID) is ignored. https://xmidt.io/docs/wrp/basics/#locators
+var LocatorPattern = regexp.MustCompile(
+	`^(?P<scheme>(?i)` + macPrefix + `|` + uuidPrefix + `|` + eventPrefix + `|` + dnsPrefix + `|` + serialPrefix + `):(?P<authority>[^/]+)?`,
+)
+
+// SpecValidators is a WRP validator that ensures messages are valid based on
+// each spec validator in the list. Only validates the opinionated portions of the spec.
+var SpecValidators = Validators{UTF8Validator, MessageTypeValidator, SourceValidator, DestinationValidator}
```

Comment:
Making this an exported `var` means that the value of `SpecValidators` is a global state that other packages can import and change.  In order to make this unable to be changed, we can make it a function:
```suggestion
func SpecValidators() Validator {
    return Validators{UTF8Validator, MessageTypeValidator, SourceValidator, DestinationValidator}
}
```

Is this comprehensive for the whole spec? Or just meant to be the extent for this PR? I would expect type specific spec information we would need to validate, but need to check...

Reply:
Great point!

I would say this is the extent for the PR atm. I was mainly focusing on https://xmidt.io/docs/wrp/basics for this pr and creating the new PR for the simple message types and so on.